proposals/0000-simple-constrained-type-families.rst

```diff
@@ -0,0 +1,363 @@
+------------
+
+A lack of constrained type families leaves many "obviously injective" type families unable to be declared as such, and leaves other families open to misuse. For an example of each, we will use a `Pred` family for inductive type level natural numbers and polymorphic arithmetic on type level numbers. For more examples of the former case, see § 4.1 in `the paper introducing injective type families <http://ics.p.lodz.pl/~stolarek/_media/pl:research:stolarek_peyton-jones_eisenberg_injectivity.pdf>`_.
+
+::
+
+    data Nat = Zero | Succ Nat
+
+    type family Pred nat = pred | pred -> nat where
+        Pred (Succ nat) = nat
+
+Currently, this is an error, because even for a closed type family, GHC is unable to recognize that a type family can have a partial domain. 
+
+:: 
+
+    • Type family equation violates injectivity annotation.
+      RHS of injective type family equation is a bare type variable
+      but these LHS type and kind patterns are not bare variables: ‘'Succ nat’
+        Pred ('Succ nat) = nat -- Defined at <interactive>:4:9
+    • In the equations for closed type family ‘Pred’
+      In the type family declaration for ‘Pred’
+
+This is preferable to the second type of error, where meaningless code can be well-typed.
+
+::
+
+    class TNum k where
+        type (a :: k) + (b :: k) :: k
+        type (a :: k) - (b :: k) :: k
+        type (a :: k) * (b :: k) :: k
+        type FromInteger (a :: Integer) :: k
+
+This is accepted by GHC, but it provides an interface that leaves much to be desired. While it is not possible to write an instance for any of these type families independently of an instance for TNum k, there is nothing stopping the programmer from using these type families nonsensically.
+
+::
+
+    newtype Const t a = Const a
+
+    pairWithProduct :: a -> b -> Const (a * b) (a, b)
+    pairWithProduct a b = Const (a, b)
+
+    what :: Const (Integer * String) (Integer, String)
+    what = pairWithProduct 42 "foo"
+
+What is the meaning of ``Integer * String``? There are arguable definitions that could be used, but it is unlikely to actually be what the programmer would like to express, at least if they haven't provided an explanation for what doing arithmetic with arbitrary types is supposed to mean by writing an instance ``TNum Type``.
+
+Proposed Change Specification
+-----------------------------
+
+Formal Specification
+++++++++++++++++++++
+
+There is a very simple way to reuse the currently existing mechanisms to give the desired behavior, with four changes to current behavior:
+
+Promote typeclass dictionary constructors
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For every class declaration ``(C1 a, C2 b) => C a b c``, a new type-level data constructor is introduced ``CDict :: C1 a -> C2 b -> C a b c``. That is, the type-level data constructor produces a type of kind ``C a b c``, taking dictionaries of any superclasses as arguments. Nothing changes if ``C`` does or does not have methods.
```

Comment:
Are non-promoted dictionaries currently currried like that today?

Reply:
Yes.

Reply:
```haskell
$fCIntBoolMaybe = \ @ a_a1E4 -> C:C $fC1Int $fC2Bool
```

Reply:
Thanks for letting me know!